### PR TITLE
bwd pass

### PR DESCRIPTION
**Pull request by liangel-02**
**Summary**
This implements the backward pass for the Varlen API and registers `_varlen_attn()` as a custom op.

**Benchmarking**

To benchmark, we compare runtime and TFLOPs against the current SDPA approach with padding.

Settings: 

- 1 H100 machine
- `batch_size=8`, `max_seq_len=2048`, `embed_dim=1024`, `num_heads=16`
- dtype `torch.bfloat16`
- `is_causal=False`
- for variable length, we set sequences to be random multiples of 64 up to `max_seq_len`
- 100 runs 

|        | Variable Length API | SDPA     |
|--------|--------------------|----------|
| Runtime | 0.8189142608642578 ms       | 3.263883056640625 ms  |
| TFLOPs | 268.652       | 158.731  |

We can see that runtime for Varlen is >3x faster

**Testing**

Run `python test/test_varlen_attention.py` for unit tests where we verify basic functionality and confirm numerical match between varlen gradients vs SDPA.

For custom op testing, `test_custom_op_registration` uses logging mode to verify that `_varlen_attn()` was called and tests with `torch.compile`. `test_custom_op_compliances` uses `torch.library.opcheck()` to verify.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164950
* __->__ #164504

### CODE REVIEW

File: torch/nn/attention/__init__.py [hunk 1]
```diff
@@ -14,14 +14,11 @@
...
 )
 
-from .varlen import varlen_attn
-
 
 __all__: list[str] = [
     "SDPBackend",
     "sdpa_kernel",
     "WARN_FOR_UNFUSED_KERNELS",
-    "varlen_attn",
```
Comment (drisspg) [inline torch/nn/attention/__init__.py:24]: I think is fine for now but curious if this was intentional

Comment (liangel-02) [reply torch/nn/attention/__init__.py:24]: yeah i had to remove this to prevent circular imports when registering as a custom op

File: test/test_varlen_attention.py [hunk 2]
```diff
@@ -147,6 +191,109 @@ def test_basic_functionality(self, device, dtype):
...
+
+        out, lse, rng_state = torch.ops.torch_nn_attention._varlen_attn(
+            q, k, v, cu_seq, cu_seq, shape.max_seq_len, shape.max_seq_len, False
+        )
+        grad_out = torch.randn_like(out)
+
+        torch.library.opcheck(
+            torch.ops.torch_nn_attention._varlen_attn_backward,
+            (grad_out, q, k, v, out, lse, cu_seq, cu_seq, shape.max_seq_len, shape.max_seq_len, False),
+            test_utils=["test_schema", "test_faketensor"] # doesn't support double bwd
```
Comment (drisspg) [inline test/test_varlen_attention.py]: hmm is there a way to disable only these tests w/ opcheck?

File: torch/nn/attention/varlen.py [hunk 3]
```diff
@@ -33,8 +33,7 @@ class AuxRequest(NamedTuple):
     lse: bool = False
 
 
-# import failures when I try to register as custom op
-# @torch.library.custom_op("torch_nn_attention::_varlen_attn", mutates_args={})
+@torch.library.custom_op("torch_nn_attention::_varlen_attn", mutates_args={})
```
Comment (drisspg) [inline torch/nn/attention/varlen.py]: lets make the lib name better, torch:: ? 

@zou3519 Is there any naming convention for python ops in PT?

File: torch/nn/attention/varlen.py [hunk 4]
```diff
@@ -86,10 +85,11 @@ def _varlen_attn(
             return_debug_mask=False,
         )
 
-    return output, softmax_lse
+    rng_state_ = torch.zeros((2,), dtype=torch.uint64, device=query.device)  # hardcoded
+    return output, softmax_lse, rng_state_
```
Comment (drisspg) [inline torch/nn/attention/varlen.py:91]: write a note somewhere about this that this is intentional for X

File: torch/nn/attention/varlen.py [hunk 5]
```diff
@@ -191,9 +193,138 @@ def varlen_attn(
...
+    ctx.key = key
+    ctx.value = value
+    ctx.cu_seq_q = cu_seq_q
+    ctx.cu_seq_k = cu_seq_k
+    ctx.max_q = max_q
+    ctx.max_k = max_k
+    ctx.is_causal = is_causal
+    ctx.output = out
+    ctx.lse = lse
+    ctx.rng_state = rng_state
```
Comment (drisspg) [inline torch/nn/attention/varlen.py]: hmm we shud be able to still passs in the dummy rng we created in the fwd?

File: torch/nn/attention/varlen.py (hunk 3 above)
Comment (drisspg) [reply torch/nn/attention/varlen.py]: @liangel-02 lets just change to `torch_attn`

**Review by drisspg — APPROVED** [review]

**Comment by amdfaa** [comment]
This PR is failing both on ROCm and CUDA. See here:
https://hud.pytorch.org/failure?name=rocm-mi300%20%2F%20linux-noble-rocm-py3.12-mi300%20%2F%20test%20(default%2C%201%2C%206%2C%20linux.rocm.gpu.gfx942.1)&jobName=undefined&failureCaptures=test%2Ftest_varlen_attention.py%3A%3ATestVarlenAttentionCUDA%3A%3Atest_varlen_vs_sdpa_float16_is_causal_True_cuda_float16
